src/liboslexec/batched_analysis.cpp

```diff
@@ -2008,7 +2008,13 @@ struct Analyzer {
     {
         bool previously_was_uniform = symbol_to_be_varying->is_uniform();
         if (previously_was_uniform | force) {
-            symbol_to_be_varying->make_varying();
+            // Interactive params are never varying, following getLLVMSymbolBase()
+            const bool force_uniform = symbol_to_be_varying->symtype()
+                                           == SymTypeParam
+                                       && symbol_to_be_varying->interactive();
+            if (!force_uniform) {
+                symbol_to_be_varying->make_varying();
```

Comment:
While it is technically correct that interactive params are always uniform, we shouldn't have to force it.
The need for this change show's that there is bug elsewhere.

Previously there was just `Symbol::lockgeom()` and that changed to `Symbol::interpolated()` and `Symbol::interactive()` and the lockgeom covered both:

```
bool Symbol::lockgeom() const
{
    // We can lock a value to a constant (at all places on all pieces of
    // geometry) if it is neither interpolated nor interactively modified.
    return !m_interpolated && !m_interactive;
}
```
However not all code that used lockgeom got updated, in particular the method:

```
void Analyzer::make_interpolated_parameters_varying()
{
    // Mark all interpolated parameters as varying,
    // As we expect interpolated data,  get_userdata will
    // only provide varying values, so we must mark
    // our symbols appropriately!
    FOREACH_PARAM(Symbol & s, inst())
    {
        if (s.everread() && !s.lockgeom() && !s.typespec().is_closure()) {
            recursively_mark_varying(&s);
        }
    }
}
```
Could you try changing it from !lockgeom() to interactive() there:

```
void Analyzer::make_interpolated_parameters_varying()
{
    // Mark all interpolated parameters as varying,
    // As we expect interpolated data,  get_userdata will
    // only provide varying values, so we must mark
    // our symbols appropriately!
    FOREACH_PARAM(Symbol & s, inst())
    {
        if (s.everread() && s.interpolated() && !s.typespec().is_closure()) {
            recursively_mark_varying(&s);
        }
    }
}
```
Then replace your force_uniform change with an
`OSL_ASSERT(!symbol_to_be_varying->interactive());`

Reply:
Might be more broken there, let me look a bit wider in scope.

Reply:
This might not be it, but was one other spot using lockgeom, in Analyzer::discover_init_symbols
```
            // Skip if it's an interpolated (userdata) parameter and we're
            // initializing them lazily.
            if ((s.symtype() == SymTypeParam
                 || s.symtype() == SymTypeOutputParam)
                && !s.lockgeom() && !s.typespec().is_closure() && !s.connected()
                && !s.connected_down() && m_ba.shadingsys().lazy_userdata())
                continue;

```                
Try changing it to more closely match code from BatchedBackendLLVM::build_llvm_instance
```
        // Skip if it's an interpolated (userdata) parameter and we're
        // initializing them lazily, or if it's an interactively-adjusted
        // parameter.
        if ((s.symtype() == SymTypeParam
                 || s.symtype() == SymTypeOutputParam)
            && !s.typespec().is_closure()
            && !s.connected() && !s.connected_down()
            && (s.interactive()
                || (s.interpolated() && shadingsys().lazy_userdata())))
            continue;

```
Note there is still a difference between the build_llvm_instance logic and the Analyzer.  As written the Analyzer allow OutputParams to be interactive or interpolated, while build_llvm_instance doesn't.  In a recent discussion, since oslc allows interpolated attribute on output params, it must be legal.  So in a different PR, I think we are changing the build_llvm_instance and llvm_instance to match.

And based on all that, I suppose the requested
`OSL_ASSERT(!symbol_to_be_varying->interactive());`
was way to simple.

It should be more like:
`OSL_ASSERT(!symbol_to_be_varying->interactive() || s.typespec().is_closure() || s.connected() || s.connected_down());`

I guess if that assert fires, I'd be interested in knowing all the symbol's data member values.

Reply:
This didn't work either.

What happens is that the `reparam` test has a parameter `float user = 2 [[ int interpolated = 1 ]]`, that is made interactive. `Analyzer::make_interpolated_parameters_varying` encounters this symbol, which is marked both interpolated and interactive, and marks it as varying.

The follwing makes all the batched tests pass. It's basically just moving the check to force interactive parameters as uniform one level up in the callstack.
```
diff --git a/src/liboslexec/batched_analysis.cpp b/src/liboslexec/batched_analysis.cpp
index 18f0edd..e46fa74 100644
--- a/src/liboslexec/batched_analysis.cpp
+++ b/src/liboslexec/batched_analysis.cpp
@@ -2303,7 +2298,8 @@ struct Analyzer {
         // our symbols appropriately!
         FOREACH_PARAM(Symbol & s, inst())
         {
-            if (s.everread() && !s.lockgeom() && !s.typespec().is_closure()) {
+            if (s.everread() && s.interpolated() && !s.interactive()
+                && !s.typespec().is_closure()) {
                 recursively_mark_varying(&s);
             }
         }
```

My understanding from reading the code is that interactive always override interpolated values. If a symbol is interactive it is not interpolated. Because the interpolated values are bound in `BackendLLVM::llvm_assign_initial_value`, but `BackendLLVM::build_llvm_instance` skips calling that for interactive parameters.

Reply:
I added a fix along these lines now.